hal/armv7m/mpu.c

```diff
@@ -20,99 +20,181 @@
 static mpu_common_t mpu_common;
 
 
+/* clang-format off */
 enum { mpu_type, mpu_ctrl, mpu_rnr, mpu_rbar, mpu_rasr, mpu_rbar_a1, mpu_rasr_a1, mpu_rbar_a2, mpu_rasr_a2,
 	   mpu_rbar_a3, mpu_rasr_a3 };
+/* clang-format on */
 
 
-/* Binary count trailing zeros */
-__attribute__((always_inline)) static inline u8 _ctz(u32 val)
+/* Removes all RASR attribute bits except ENABLE */
+#define HOLE_ATTR(rasrAttr) (0 | ((rasrAttr) & 0x1))
+
+
+/* Setup single MPU region entry in a local MPU context */
+static int mpu_regionSet(unsigned int *idx, u32 base_addr, u8 srdMask, u8 sizeBit, u32 rasrAttr)
 {
-	asm volatile(" \
-		rbit %0, %0; \
-		clz %0, %0"
-		: "+l"(val));
-	return val;
+	if ((sizeBit < 5) || (*idx >= mpu_common.regMax)) {
+		return -EPERM;
+	}
+
+	mpu_common.region[*idx].rbar =
+			base_addr |
+			(1ul << 4) | /* mark region as valid */
+			(*idx & 0xful);
+
+	mpu_common.region[*idx].rasr =
+			rasrAttr |
+			(((u32)srdMask) << 8) |
+			((((u32)sizeBit - 1) & 0x1ful) << 1);
+
+	*idx += 1;
+	return EOK;
 }
 
 
-/* Setup single MPU region entry in a local MPU context */
-static int mpu_regionSet(u8 idx, addr_t addr, u8 sizeBit, u32 attr, unsigned int enable, u8 srdMask)
+/* Translate memory map attributes to RASR attribute bits */
+static u32 mpu_regionAttrs(u32 attr, unsigned int enable)
 {
 	u8 tex = 0;
 	u8 ap = 1; /* privileged read-write access, unprivileged no access */
 
-	if (sizeBit < 5 || idx >= mpu_common.regMax)
-		return -EPERM;
-
-	if (attr & mAttrRead)
+	if ((attr & mAttrRead) != 0) {
 		ap = 2; /* privileged read-write access, unprivileged read only access */
+	}
 
-	if (attr & mAttrWrite)
+	if ((attr & mAttrWrite) != 0) {
 		ap = 3; /* privileged read-write access, unprivileged read and write access */
+	}
 
-	mpu_common.region[idx].rbar =
-		((u32)addr & (0x7fffffful << 5)) |
-		(1ul << 4) | /* mark region as valid */
-		(idx & 0xful);
-
-	mpu_common.region[idx].rasr =
-		((((attr & mAttrExec) == 0) & 1ul) << 28) |
-		((ap & 0x7ul) << 24) |
-		((tex & 0x7ul) << 19) |
-		((((attr & mAttrShareable) != 0) & 1ul) << 18) |
-		((((attr & mAttrCacheable) != 0) & 1ul) << 17) |
-		((((attr & mAttrBufferable) != 0) & 1ul) << 16) |
-		((srdMask & 0xfful) << 8) |
-		(((sizeBit - 1) & 0x1ful) << 1) |
-		(enable != 0);
-
-	return EOK;
+	return ((((attr & mAttrExec) == 0) & 1ul) << 28) |
+			((ap & 0x7ul) << 24) |
+			((tex & 0x7ul) << 19) |
+			((((attr & mAttrShareable) != 0) & 1ul) << 18) |
+			((((attr & mAttrCacheable) != 0) & 1ul) << 17) |
+			((((attr & mAttrBufferable) != 0) & 1ul) << 16) |
+			(enable != 0);
 }
 
 
-/* Find best alignment for map and setup a region */
-static int mpu_regionBestFit(u8 idx, addr_t addr, size_t size, u32 attr, unsigned int enable, size_t *allocSize)
+static int mpu_checkOverlap(unsigned int idx, u32 start, u32 end)
 {
-	size_t srSize;
-	addr_t srBase, alignMask;
+	end -= 1;
+	for (int i = 0; i < idx; i++) {
```

Comment:
Not allowed in kernel and plo, please move variables declarations to the beginning of the function

Reply:
OK, moved `const` declarations to the start as well, but I'm not sure if it was necessary.